state-chain/pallets/cf-elections/src/electoral_systems/solana_swap_accounts_tracking.rs

```diff
@@ -0,0 +1,274 @@
+		E: sp_std::fmt::Debug + 'static,
+		Account: MaybeSerializeDeserialize + Member + Parameter + Ord,
+		SwapDetails: MaybeSerializeDeserialize + Member + Parameter + Ord,
+		BlockNumber: MaybeSerializeDeserialize + Member + Parameter + Ord + CheckedSub + Into<u32> + Copy,
+		Settings: Member + Parameter + MaybeSerializeDeserialize + Eq,
+		Hook: SolanaVaultSwapAccountsHook<Account, SwapDetails, E> + 'static,
+		ValidatorId: Member + Parameter + Ord + MaybeSerializeDeserialize,
+	> ElectoralSystem
+	for SolanaVaultSwapAccounts<Account, SwapDetails, BlockNumber, Settings, Hook, ValidatorId, E>
+{
+	type ValidatorId = ValidatorId;
+	type ElectoralUnsynchronisedState = SolanaVaultSwapAccountsLastClosedAt<BlockNumber>;
+	type ElectoralUnsynchronisedStateMapKey = ();
+	type ElectoralUnsynchronisedStateMapValue = ();
+
+	type ElectoralUnsynchronisedSettings = ();
+	type ElectoralSettings = Settings;
+	type ElectionIdentifierExtra = ();
+	type ElectionProperties = SolanaVaultSwapsKnownAccounts<Account>;
+	type ElectionState = ();
+	type Vote = vote_storage::bitmap::Bitmap<SolanaVaultSwapsVote<Account, SwapDetails>>;
+	type Consensus = SolanaVaultSwapsVote<Account, SwapDetails>;
+	type OnFinalizeContext = BlockNumber;
+	type OnFinalizeReturn = ();
+
+	fn generate_vote_properties(
+		_election_identifier: ElectionIdentifier<Self::ElectionIdentifierExtra>,
+		_previous_vote: Option<(VotePropertiesOf<Self>, AuthorityVoteOf<Self>)>,
+		_vote: &<Self::Vote as VoteStorage>::PartialVote,
+	) -> Result<VotePropertiesOf<Self>, CorruptStorageError> {
+		Ok(())
+	}
+
+	fn is_vote_desired<ElectionAccess: ElectionReadAccess<ElectoralSystem = Self>>(
+		_election_identifier_with_extra: crate::electoral_system::ElectionIdentifierOf<Self>,
+		_election_access: &ElectionAccess,
+		_current_vote: Option<(VotePropertiesOf<Self>, AuthorityVoteOf<Self>)>,
+	) -> Result<bool, CorruptStorageError> {
+		Ok(true)
+	}
+
+	fn is_vote_needed(
+		(_, current_partial_vote, _): (
+			VotePropertiesOf<Self>,
+			<Self::Vote as VoteStorage>::PartialVote,
+			AuthorityVoteOf<Self>,
+		),
+		(proposed_partial_vote, _): (
+			<Self::Vote as VoteStorage>::PartialVote,
+			<Self::Vote as VoteStorage>::Vote,
+		),
+	) -> bool {
+		current_partial_vote != proposed_partial_vote
+	}
+
+	fn on_finalize<ElectoralAccess: ElectoralWriteAccess<ElectoralSystem = Self>>(
+		electoral_access: &mut ElectoralAccess,
+		election_identifiers: Vec<ElectionIdentifier<Self::ElectionIdentifierExtra>>,
+		current_block_number: &Self::OnFinalizeContext,
+	) -> Result<Self::OnFinalizeReturn, CorruptStorageError> {
+		if let Some(election_identifier) = election_identifiers
+			.into_iter()
+			.at_most_one()
+			.map_err(|_| CorruptStorageError::new())?
+		{
+			let mut election_access = electoral_access.election_mut(election_identifier)?;
+			if let Some(consensus) = election_access.check_consensus()?.has_consensus() {
+				let mut known_accounts = election_access.properties()?;
+				election_access.delete();
```

Comment:
I would move this to right before new_election, just so it's easier to see the fact we only expect (and always expect) one election to open at any one time.

Reply:
this is not possible i guess since the rust compiler does not allow you to borrow immutably while we have already borrowed the same thing mutably. We have that kind of a case here: we have to call the ```				election_access.delete();``` before calling ```electoral_access.unsynchronised_state()```

Reply:
interesting... might be possible after my refactor - but ok to leave it for now, is a tiny thing